### DISCUSSION TITLE
Property ScoutEngines\Postgres\PostgresEngine::$database (Illuminate\Database\Connection) does not accept Illuminate\Database\ConnectionInterface.

### DISCUSSION THREAD

**Discussion by devNoiseConsulting** [larastan/larastan#1564] [Q&A] [answered]
I've forked a package (pmatseykanets/laravel-scout-postgres) and updating it for the new version of Laravel. While doing that I'm running Larastan on the code to clean it up a bit. I'm a bit stumped on how to fix the following error:

`Property ScoutEngines\Postgres\PostgresEngine::$database (Illuminate\Database\Connection) does not accept Illuminate\Database\ConnectionInterface.`

The code is trying to get the database connection through the `$this->resolver` which is `Illuminate\Database\ConnectionResolverInterface` and the `connection()` method returns a `Illuminate\Database\ConnectionInterface`.

The `$database` variable is cast as `Illuminate\Database\Connection`. Which is an implemenation of `Illuminate\Database\ConnectionInterface`.

I understand that larastan/phpstan doesn't like that I'm assigning the `$connection` to the `$database` because it thinks there is a type mismatch. I would change the type of `$database` to `Illuminate\Database\ConnectionInterface`, but the code is making use of methods in `Illuminate\Database\Connection` that are not present in `Illuminate\Database\ConnectionInterface`. 

Is there any way to cast `$connection` to a `Illuminate\Database\Connection`? Or do I just need to make an exception for this assignment? 

The relavent code:
```php
    /**
     * Database connection.
     *
     * @var \Illuminate\Database\Connection
     */
    protected $database;

    /**
     * Database connection resolver.
     *
     * @var \Illuminate\Database\ConnectionResolverInterface
     */
    protected $resolver;

    /**
     * Connect to the database.
     * @return void
     */
    protected function connect()
    {
        // Already connected
        if ($this->database !== null) {
            return;
        }

        $connection = $this->resolver->connection($this->config('connection'));

        if ($connection->getDriverName() !== 'pgsql') {
            throw new \InvalidArgumentException('Connection should use pgsql driver.');
        }

        $this->database = $connection;
    }
```

**Comment by szepeviktor** [comment, marked as answer]
Hello!

> Is there any way to cast

Yes, add `/** @var \Illuminate\Database\Connection $connection */` above `$connection =`

**Reply by devNoiseConsulting** [reply]
Thank you. That eliminated the error. Still getting up to speed how to use larastan.

**Comment by szepeviktor** [comment]
BTW consider using positive conditions.
`if ($this->database !== null) {`

**Reply by devNoiseConsulting** [reply]
Yeah, looking at the code with that feedback makes sense. Gonna hold off on making that change, but it's on the todo list.

**Reply by szepeviktor** [reply]
🖱️ https://github.com/szepeviktor/very-clean-code
